src/Pate/Verification/AbstractDomain.hs

```diff
@@ -0,0 +1,506 @@
+
+widenAbsDomainVals' ::
+  forall sym arch bin m.
+  Monad m =>
+  IO.MonadIO m =>
+  W4.IsSymExprBuilder sym =>
+  PA.ValidArch arch =>
+  sym ->
+  AbstractDomainVals sym arch bin {- ^ existing abstract domain that this is augmenting -} ->
+  (forall tp. W4.SymExpr sym tp -> m (AbsRange tp)) ->
+  PS.SimOutput sym arch bin ->
+  m (AbstractDomainVals sym arch bin, WidenLocs sym arch)
+widenAbsDomainVals' sym prev f stOut = CMW.runWriterT $ do
+  memVals <- MapF.traverseMaybeWithKey relaxMemVal (absMemVals prev)
+  regVals <- PRt.zipWithRegStatesM (absRegVals prev) (PS.simOutRegs stOut) relaxRegVal
+  return $ AbstractDomainVals regVals memVals
+
+  where
+    getMemAbsVal ::
+      PMC.MemCell sym arch w ->
+      m (MemAbstractValue sym w)
+    getMemAbsVal cell = do
+      val <- IO.liftIO $ PMC.readMemCell sym (PS.simOutMem stOut) cell
+      MemAbstractValue <$> getAbsVal sym f (PSR.ptrToEntry val)
+
+    relaxMemVal ::
+      PMC.MemCell sym arch w ->
+      MemAbstractValue sym w {- ^ previous abstract value -} ->
+      CMW.WriterT (WidenLocs sym arch) m (Maybe (MemAbstractValue sym w))
+    relaxMemVal cell (MemAbstractValue absValPrev) = case isUnconstrained absValPrev of
+      True -> return Nothing
+      False -> do
+        MemAbstractValue absValNew <- CMW.lift $ getMemAbsVal cell
+        let absValCombined = combineAbsVals absValNew absValPrev
+        unless (absValCombined == absValPrev) $ CMW.tell (WidenLocs mempty (S.singleton (Some cell)))
+        case isUnconstrained absValCombined of
+          True -> return Nothing
+          False -> return $ Just $ MemAbstractValue absValCombined
+        
+    relaxRegVal ::
+      MM.ArchReg arch tp ->
+      MacawAbstractValue sym tp {- ^ previous abstract value -} ->
+      PSR.MacawRegEntry sym tp ->
+      CMW.WriterT (WidenLocs sym arch) m (MacawAbstractValue sym tp)
+    relaxRegVal reg absValPrev v = case isUnconstrained absValPrev of
+      True -> return absValPrev
+      False -> do
+        absValNew <- CMW.lift $ getAbsVal sym f v
+        let absValCombined = combineAbsVals absValNew absValPrev
+        unless (absValCombined == absValPrev) $ CMW.tell (WidenLocs (S.singleton (Some reg)) mempty)
+        return absValCombined
+
+widenAbsDomainVals ::
+  forall sym arch m.
+  Monad m =>
+  IO.MonadIO m =>
+  W4.IsSymExprBuilder sym =>
+  PA.ValidArch arch =>
+  sym ->
+  AbstractDomainBody sym arch {- ^ existing abstract domain that this is augmenting -} ->
+  (forall tp. W4.SymExpr sym tp -> m (AbsRange tp)) ->
+  PS.SimBundle sym arch ->
+  m (AbstractDomainBody sym arch, Maybe (WidenLocs sym arch))
+widenAbsDomainVals sym prev f bundle = do
+  let (PPa.PatchPair absValsO absValsP) = absDomVals prev
+  (absValsO', locsO) <- widenAbsDomainVals' sym absValsO f (PS.simOutO bundle)
+  (absValsP', locsP) <- widenAbsDomainVals' sym absValsP f (PS.simOutP bundle)
+  return $ (prev { absDomVals = PPa.PatchPair absValsO' absValsP' }, Just $ locsO <> locsP)
```

Comment:
I think it would be worth describing the strategy here for readers who are not familiar with constant propagation.